File: app/src/main/java/com/example/purepawapp/ui/admin/AdminBlogsFragment.java
package com.example.purepawapp.ui.admin;

import android.os.Bundle;
import android.view.View;

import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;

import com.example.purepawapp.R;
import com.example.purepawapp.data.model.BlogPost;
import com.example.purepawapp.data.repository.RepoCallback;
import com.example.purepawapp.databinding.FragmentAdminBlogsBinding;
import com.example.purepawapp.di.ServiceLocator;
import com.example.purepawapp.ui.common.BaseFragment;
import com.example.purepawapp.util.SpacingItemDecoration;
import com.example.purepawapp.util.ViewUtils;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import java.util.ArrayList;
import java.util.List;

public class AdminBlogsFragment extends BaseFragment<FragmentAdminBlogsBinding> {

    private List<BlogPost> allBlogs = List.of();

    private final AdminBlogAdapter adapter = new AdminBlogAdapter(blog ->
            new MaterialAlertDialogBuilder(requireContext())
                    .setTitle("Xóa bài viết")
                    .setMessage("Bạn có chắc muốn xóa \"" + blog.getTitle() + "\"?")
                    .setNegativeButton("Hủy", null)
                    .setPositiveButton("Xóa", (dialog, which) -> deleteBlog(blog.getId()))
                    .show());

    public AdminBlogsFragment() {
        super(FragmentAdminBlogsBinding::inflate);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        getBinding().btnBack.setOnClickListener(v -> NavHostFragment.findNavController(this).popBackStack());
        getBinding().rvBlogs.setLayoutManager(new LinearLayoutManager(requireContext()));
        getBinding().rvBlogs.setAdapter(adapter);
        getBinding().rvBlogs.addItemDecoration(new SpacingItemDecoration((int) (12 * getResources().getDisplayMetrics().density)));

        getBinding().btnAddBlog.setOnClickListener(v -> NavHostFragment.findNavController(this).navigate(R.id.action_adminBlogsFragment_to_adminAddBlogFragment));
        getBinding().chipGroupFilter.setOnCheckedStateChangeListener((group, checkedIds) -> applyFilter());

        loadBlogs();
    }

    @Override
    public void onResume() {
        super.onResume();
        loadBlogs();
    }

    private void loadBlogs() {
        ServiceLocator.getBlogRepository().getBlogs(new RepoCallback<>() {
            @Override
            public void onSuccess(List<BlogPost> result) {
                allBlogs = result;
                applyFilter();
            }

            @Override
            public void onError(Exception error) {
                ViewUtils.toast(AdminBlogsFragment.this, error.getMessage() != null ? error.getMessage() : "Không thể tải danh sách blog");
            }
        });
    }

    private void applyFilter() {
        int checkedId = getBinding().chipGroupFilter.getCheckedChipId();
        List<BlogPost> filtered;
        if (checkedId == getBinding().chipPublished.getId()) {
            filtered = filterByStatus("published");
        } else if (checkedId == getBinding().chipDraft.getId()) {
            filtered = filterByStatus("draft");
        } else {
            filtered = allBlogs;
        }
        adapter.submitList(filtered);
        getBinding().emptyState.setVisibility(filtered.isEmpty() ? View.VISIBLE : View.GONE);
    }

    private List<BlogPost> filterByStatus(String status) {
        List<BlogPost> result = new ArrayList<>();
        for (BlogPost b : allBlogs) if (status.equals(b.getStatus())) result.add(b);
        return result;
    }

    private void deleteBlog(String blogId) {
        ServiceLocator.getBlogRepository().deleteBlog(blogId, new RepoCallback<>() {
            @Override
            public void onSuccess(Void result) {
                ViewUtils.toast(AdminBlogsFragment.this, "Đã xóa bài viết");
                loadBlogs();
            }

            @Override
            public void onError(Exception error) {
                ViewUtils.toast(AdminBlogsFragment.this, error.getMessage() != null ? error.getMessage() : "Không thể xóa bài viết");
            }
        });
    }
}

File: app/src/main/java/com/example/purepawapp/ui/admin/AdminBlogsFragment.kt
package com.example.purepawapp.ui.admin

import android.os.Bundle
import android.view.View
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.purepawapp.R
import com.example.purepawapp.data.model.BlogPost
import com.example.purepawapp.databinding.FragmentAdminBlogsBinding
import com.example.purepawapp.di.ServiceLocator
import com.example.purepawapp.ui.common.BaseFragment
import com.example.purepawapp.util.SpacingItemDecoration
import com.example.purepawapp.util.toast
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import kotlinx.coroutines.launch

class AdminBlogsFragment : BaseFragment<FragmentAdminBlogsBinding>(FragmentAdminBlogsBinding::inflate) {

    private var allBlogs: List<BlogPost> = emptyList()

    private val adapter = AdminBlogAdapter(
        onDelete = { blog ->
            MaterialAlertDialogBuilder(requireContext())
                .setTitle("Xóa bài viết")
                .setMessage("Bạn có chắc muốn xóa \"${blog.title}\"?")
                .setNegativeButton("Hủy", null)
                .setPositiveButton("Xóa") { _, _ -> deleteBlog(blog.id) }
                .show()
        }
    )

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding.btnBack.setOnClickListener { findNavController().popBackStack() }
        binding.rvBlogs.layoutManager = LinearLayoutManager(requireContext())
        binding.rvBlogs.adapter = adapter
        binding.rvBlogs.addItemDecoration(SpacingItemDecoration((12 * resources.displayMetrics.density).toInt()))

        binding.btnAddBlog.setOnClickListener {
            findNavController().navigate(R.id.action_adminBlogsFragment_to_adminAddBlogFragment)
        }
        binding.chipGroupFilter.setOnCheckedStateChangeListener { _, _ -> applyFilter() }

        loadBlogs()
    }

    override fun onResume() {
        super.onResume()
        loadBlogs()
    }

    private fun loadBlogs() {
        viewLifecycleOwner.lifecycleScope.launch {
            ServiceLocator.blogRepository.getBlogs()
                .onSuccess {
                    allBlogs = it
                    applyFilter()
                }
                .onFailure { toast(it.message ?: "Không thể tải danh sách blog") }
        }
    }

    private fun applyFilter() {
        val filtered = when (binding.chipGroupFilter.checkedChipId) {
            binding.chipPublished.id -> allBlogs.filter { it.status == "published" }
            binding.chipDraft.id -> allBlogs.filter { it.status == "draft" }
            else -> allBlogs
        }
        adapter.submitList(filtered)
        binding.emptyState.visibility = if (filtered.isEmpty()) View.VISIBLE else View.GONE
    }

    private fun deleteBlog(blogId: String) {
        viewLifecycleOwner.lifecycleScope.launch {
            ServiceLocator.blogRepository.deleteBlog(blogId)
                .onSuccess {
                    toast("Đã xóa bài viết")
                    loadBlogs()
                }
                .onFailure { toast(it.message ?: "Không thể xóa bài viết") }
        }
    }
}
